main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright 2022 Anna Krogh, Patrick Lundvig, Christian Bonde, 2025 Jakob Schneider Villumsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur.*
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.{DefContext, Occur}
+import ca.uwaterloo.flix.language.ast.Symbol.{DefnSym, VarSym}
+import ca.uwaterloo.flix.language.ast.{OccurrenceAst, Symbol}
+import ca.uwaterloo.flix.util.ParOps
+
+/**
+  * The occurrence analyzer collects information on variable and function usage and calculates the weight of the expressions
+  * Marks a variable or function as Dead if it is not used, Once if it is used exactly once and Many otherwise
+  */
+object OccurrenceAnalyzer {
+
+  private object OccurInfo {
+
+    /**
+      * Occurrence information for an empty sequence of expressions
+      */
+    val Empty: OccurInfo = OccurInfo(Map.empty, Map.empty, 0, 0)
+
+    /**
+      * The initial occurrence information for an expression of size 1, i.e. an expression without subexpressions.
+      */
+    val One: OccurInfo = OccurInfo(Map.empty, Map.empty, 0, 1)
+  }
+
+  /**
+    * The occurrence of `defs` and `vars` inside the body of a `def`
+    * `size` represents the number of expressions in the body of a `def`
+    */
+  case class OccurInfo(defs: Map[DefnSym, Occur], vars: Map[VarSym, Occur], localDefs: Int, size: Int) {
+    def :+(kv: (DefnSym, Occur)): OccurInfo = {
+      this.copy(defs = this.defs + kv)
+    }
+
+    def +(kv: (VarSym, Occur)): OccurInfo = {
```

Comment:
I don't like that we have both :+ and +.

How much do we need?

Reply:
This is because `+` cannot be overloaded so we need two distinct symbols

Reply:
But thats terrible design no? Two different symbols?

Reply:
Maybe just an overloaded `add` function?